server/src/storage/segment.rs

```diff
@@ -0,0 +1,187 @@
+/* Copyright 2022 The MiniModelarDB Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! The main SegmentBuilder struct provides support for inserting and storing data in a in-memory
+//! segment. Furthermore, the data can be retrieved as a structured record batch.
+
+use crate::storage::data_point::DataPoint;
+use crate::storage::{MetaData, Timestamp, INITIAL_BUILDER_CAPACITY};
+use datafusion::arrow::array::{ArrayBuilder, Float32Builder, TimestampMicrosecondBuilder};
+use datafusion::arrow::datatypes::TimeUnit::Microsecond;
+use datafusion::arrow::datatypes::{DataType, Field, Schema};
+use datafusion::arrow::record_batch::RecordBatch;
+use datafusion::parquet::arrow::ArrowWriter;
+use datafusion::parquet::basic::Encoding;
+use datafusion::parquet::file::properties::WriterProperties;
+use std::fmt::Formatter;
+use std::fs::File;
+use std::sync::Arc;
+use std::{fmt, fs};
+
+/// A single segment being built, consisting of a series of timestamps and values. Note that
+/// since array builders are used, the data can only be read once the builders are finished and
+/// can not be further appended to after.
+pub struct SegmentBuilder {
+    /// Builder consisting of timestamps with microsecond precision.
+    timestamps: TimestampMicrosecondBuilder,
+    /// Builder consisting of float values.
+    values: Float32Builder,
+    /// Metadata used to uniquely identify the segment (and related sensor).
+    pub metadata: MetaData,
+    /// First timestamp used to distinguish between segments from the same sensor.
+    first_timestamp: Timestamp,
+}
+
+impl fmt::Display for SegmentBuilder {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str(&*format!("Segment with {} data point(s) (", self.timestamps.len()));
+        f.write_str(&*format!("timestamp capacity: {}, ", self.timestamps.capacity()));
+        f.write_str(&*format!("values capacity: {})", self.values.capacity()));
+
+        Ok(())
+    }
+}
+
+impl SegmentBuilder {
+    pub fn new(data_point: &DataPoint) -> Self {
```

Comment:
Since `metadata` and `first_timestamp` were only used for saving the compressed data, these have been removed from `SegmentBuilder` in the next branch. This is instead handled in the storage engine.